Use relative instead of absolute link in wCMS::url(); Change to cURL for grabbing version number.

Alamantus opened the pull request:
When using `wCMS::url()` as delivered on my web host, it was returning insanely long urls because of how my host's server is set up. Since all pages are displayed on the same level and because `wCMS::url()` is only used for system/menu navigation, switching this from a full url that relies on `$_SERVER['DOCUMENT_ROOT']` to a simple relative `./` link allows the method to work appropriately no matter where WonderCMS is installed.

Also, my server was never able to grab the version number from `https://raw.githubusercontent.com/robiso/wondercms/master/version` because it is cross-domain _and_ cross-protocol (http-to-https), which meant that PHP's `file_get_contents()` function did not work. Adding a private `wCMS::_getOfficialVersion()` method that uses cURL to access the `version` file worked more reliably because that is cURL's intended use, while `file_get_contents()` is mostly intended for getting contents from the same domain.

robiso commented:
thank you @Alamantus for contributing to this project.

In this case, I'm torn between your solution and this one https://github.com/robiso/wondercms/issues/18
 which uses `$_SERVER['SCRIPT_NAME']`. Can we make a short discussion on which you think would be better? Your solution, or the solution as posted above.

Additional wiki for reference:
https://github.com/robiso/wondercms/wiki/URLs-mailformed-on-Windows-IIS

Additionally, mind sending me over your name/last name + website so we can get it listed on wondercms.com/special-contributors and our download page?

Alamantus commented:
When I work with URLs, I typically lean toward relative links over absolute links because absolute links force an extra http call that relative links don't. Extra http calls put unnecessary load on the server (albeit very small). Plus the relative link solution in my PR is much smaller. :)

I see what the other solution is trying to do, but I just don't think that reconstruction the full URL is beneficial when all browsers will do that for you anyway, especially considering the fact that _all_ pages get run through `index.php`. Using `./` will always make the browser give you the correct link relative to wherever `index.php` happens to be, whereas trying to rely on (in my opinion) unnecessary `str_replace()` calls on PHP's `$_SERVER` vars might not always be reliable.

What do you think?

Alamantus commented:
> Additionally, mind sending me over your name/last name + website so we can get it listed on wondercms.com/special-contributors and our download page?

Sure! My name is Robbie Antenesse, and my website is http://robbie.antenesse.net

Thanks! :)

Alamantus commented:
> absolute links force an extra http call that relative links don't. Extra http calls put unnecessary load on the server (albeit very small).

I just looked this up because I was thinking about it, and I'm wrong on this point because every page change is an http call. :)

There are however significantly fewer PHP concatenations and much better readability in my code. There's not much overhead in concatenation, but readability is always very valuable in my opinion. Also, using relative links means you don't need to go through the trouble of trying to evaluate what the domain and protocol are since the browser already knows because it's already using it.

robiso commented:
I couldn't agree more. Your clear explanation to this issue confirms this is a definite step in the right direction. Thank you for finding a solution to the problem we didn't even know existed (updating part) and for finding the solution to the relative URL function.

Your details will be published in the next 24 hours. Could you please do an (URL) test with any of the plugins, or do you think this change does not affect any of them?

I will additionally ask wdj to run this on his server, since he always had the same issues.
If all goes well, we'll push out 2.0.4 with your two fixes and additionally publicly thank you on our /whatsnew page + forum.

Alamantus commented:
The ultimate result _should_ be that nothing changes except that it's more reliable in any environment, but I'll definitely test it with them. Makes me glad there are so few. :smile: I'll comment again with what I discover.

wdj's problem should also be resolved/improved because his solution restricts the URL to _always_ only be at `example.com/location`, never at `example.com/subdirectory/location` the way simply using `./` would allow.

Thanks! :smiley:

Alamantus commented:
Alright, I tested all the plugins, and it looks like there are no issues with any of them!

I would like to mention, however, that the emails from the contact form plugin went directly to my spam box in Gmail, so you may want to make a note about that in the plugin notes https://github.com/robiso/contact_form/releases/tag/2.0

Alamantus commented:
Sorry, I added a third commit because I realized all the updates were being done via `file_get_contents()`, so I created a `wCMS::_getExternalFile()` method that uses cURL and replaced all external calls with that method instead. I also updated the `wCMS::_getOfficialVersion()` method to use it, just to keep it consistent and cleaner. :)

robiso commented:
@Alamantus absolutely love your work, thank you for the extensive testing and explanations. Looks like you've already fully mastered WonderCMS. :grin: 

I ran a couple of tests on our servers, flawless. The only thing I didn't test was subdomains, tested https and it works awesomely! I tested the latest commit https://raw.githubusercontent.com/Alamantus/wondercms/6deef5fb73e12d25998679586c7eba7f9ced07a2/index.php please correct me if I'm wrong or if there is a newer one.

I'll write up about the spam plugin as you suggested, thank you for reporting it.
You have been also published on:
https://www.wondercms.com/special-contributors and
https://www.wondercms.com/latest

When we push out 2.0.4, which I say should happen ASAP if there are no objections, we'll also do a _honourable_ mention on:
https://www.wondercms.com/whatsnew and
https://www.wondercms.com/community

Any objections to not pushing out 2.0.4 with @Alamantus fixes?

wdj-ac commented:
@Alamantus nice work, Thank you! I will test it with my new improvements tomorrow.

Alamantus commented:
> The only thing I didn't test was subdomains,
@robiso I've _exclusively_ tested it on subdomains, so we're good there, too! :grin:

Also, yes, that is the correct commit. :)

Thanks!

robiso commented:
I'd like to take this opportunity and thank @Alamantus and @wdj-ac - thank you guys.

Noticed a small bug in the settings panel "Login URL (save your URL: **./loginURL**)".
Is there a way we can still echo the full login URL here, so it's easier for users to bookmark?

Alamantus commented:
> Noticed a small bug in the settings panel "Login URL (save your URL: ./loginURL)".
> Is there a way we can still echo the full login URL here so it's easier for users to bookmark?

@robiso Good find! I was able to think of a couple of options, depending on how or where you want the calculation to occur:

1. Save the URI (`$_SERVER['REQUEST_URI']`) in the database on database creation.
    * This would be fairly simple, but it would force users to update the setting within the database file if they move the website to a different location.
    * The good news is, even if it is wrong, it won't affect the website's loading, it will just display misinformation (i.e. "save your URL:" would display what was listed rather than the correct location, if it was moved)

2. Calculate the full URL in a separate function for display purposes only. This is what I came up with:
    * Examines the `REQUEST_URI` and checks if the last part of it is in the config's `menuItems` list. If it is, it returns the full URL minus that piece, otherwise, it just provides the full `REQUEST_URI`. (see code below)
    * The drawback of this is if the user names the subdirectory holding WonderCMS the same name as one of the menu items (i.e. if this is one of their pages `example.com/cms/cms`), then it'll strip out the subdomain (i.e. running `wCMS::location()` at `example.com/cms` would output `http://example.com/`)

```
public static function location() {
	$path = explode('/', trim($_SERVER[REQUEST_URI], '/'));
	return (isset($_SERVER['HTTPS']) ? 'https' : 'http') . '://' . $_SERVER['HTTP_HOST'] . '/' . (in_array($path[count($path) - 1], wCMS::get('config','menuItems')) ? implode(array_slice($path, 0, count($path) - 1)) : trim($_SERVER[REQUEST_URI], '/'));
}
```

Both of these ideas obviously have minor drawbacks, but I don't think they're unreasonable to ask of the user.

What do you think?

Alamantus commented:
After testing, I found that this would be the most reliable way to write `wCMS::_getLocation()`:

```
public static function _getLocation() {
	$path = explode('/', trim($_SERVER[REQUEST_URI], '/'));
	return trim((isset($_SERVER['HTTPS']) ? 'https' : 'http') . '://' . $_SERVER['HTTP_HOST'] . '/' . (in_array($path[count($path) - 1], wCMS::get('config','menuItems')) ? implode(array_slice($path, 0, count($path) - 1)) : trim($_SERVER[REQUEST_URI], '/')), '/') . '/';
}
```

Then the string would be updated as follows:

```
... (save your URL: ' . wCMS::_getLocation() . wCMS::get('config', 'login') . ') ...
```

I'm not sure which you would prefer, though.

There's a chance that the first option would ultimately be better because it takes fewer calculations, but let me know what you think.

robiso commented:
Personally, I like the first option better, but since we need this URL display only in one place and only once, maybe we don't need an extra function and just echo the path once in place where in needs to appear? This way we get rid of the unnecessary user interaction when moving the site, etc, without actually saving the ` $_SERVER[REQUEST_URI]`

I have obviously not mastered my $_SERVER knowledge, but would something like this work?
`... (save your URL: ' . $_SERVER[REQUEST_URI] . wCMS::get('config', 'login') . ') ...`
// maybe an extra . "/" . is missing in between


Another option is to not worry about this, strip the full path in the login URL in the settings panel, but I'm afraid some users would lock themselves out by not knowing what their login URL actually is. On another side, we don't want WonderCMS to keep increasing in size and functions to retain the core "Wonder". We don't want to overcomplicate things that should be simple, but it seems inevitable in a project like this. Any opinion on this topic?

Alamantus commented:
If we go with the first option, then we'd need the extra processing in the method to generate the full URL because `$_SERVER['REQUEST_URI']` only provides the _current path_, without the domain, which in your example would produce "(save your URL: /wondercms/loginURL)" or worse "(save your URL: /wondercms/exampleloginURL)" if you're on the example page.

Using the first option would require that the database.js file be updated. I edited my post before this to show all the edits required, ultimately adding 5 new lines.

The second option would significantly lengthen the line, but it would only add 1 or 2 extra lines...

What do you think?

robiso commented:
Warned you that my $_SERVER wasn't my best. :grin:, as always, thank you for the very clear explanation

After all the clarifications, I'm more inclined to use your second solution. No database hassle, we have an extra line or two, significant length or not, that part of will be hardly modified by anyone from this point onwards, so it's a :+1: from me and we'll probably push this as 2.0.5, or maybe 2.1.0 if we include @wdj-ac commits https://github.com/wdj-ac/wondercms/blob/master/index.php which include menu visibility and order changing + hook priorities (please let me know what you think if you happen to throw an eye on this topic).

Alamantus commented:
Alright, well then I'll make a new pull request for that change! Hopefully the drawbacks don't bite us later. :sweat_smile: We might want to put a warning about naming a page the same name as the subdirectory it's in? Not totally necessary since it's 1) not likely to happen, and 2) it only affects the display in that one place.

I don't have much to say on wdj's updates other than that I think those features would be very nice to have! :) I do think that this change should be a 2.0.5 version, though, since it's a direct patch to the last one.

robiso commented:
Thank you very much Robbie, 2.0.5 pushing out as we speak! Credits also added. :+1: 

I think the most appropriate place to write this is the wiki? If you agree I'll add it there.

Alamantus commented:
Happy to help!

Yeah, I think the wiki would be the right place, since that's where other tips like that have been put in the past. :)

robiso commented:
Added to wiki https://github.com/robiso/wondercms/wiki + thanks note. You're awesome!
